### ISSUE TITLE
Datadog CI Test Runs might present different tests in the same file as if they were the same

### ISSUE THREAD

**Issue by apizarro-paypal** [DataDog/dd-trace-py#4550]
### Which version of dd-trace-py are you using?
v0.56.1, but the issue _should_ also occur in the latest version of dd-trace-py, since the potential culprit line is the same in v0.56.1 ([link](https://github.com/DataDog/dd-trace-py/blob/v0.56.1/ddtrace/contrib/pytest/plugin.py#L125)) and the `1.x` branch ([link](https://github.com/DataDog/dd-trace-py/blob/1.x/ddtrace/contrib/pytest/plugin.py#L155)).

### Which version of pip are you using?
20.3.4 (Python 3.5)

### Which libraries and their versions are you using?
I've only seen the issue in Datadog CI Test Runs results published from CI, and haven't had the chance to run `pip freeze` in the CI app container yet.

### How can we reproduce your problem?
Create a `tests/test_it.py` Python module with the following tests:

```python
import unittest

class TestFoo(unittest.TestCase):
    def test_whatever(self):
        pass


class TestBar(unittest.TestCase):
    def test_whatever(self):
        pass
```

(I'm presuming a Python project that contains tests in a separate `tests/` root folder. The `tests/` folder must be a Python package with a `tests/__init__.py` file.)

### What is the result that you get?
Datadog CI Test Runs presents results for both tests under the same Test Suite and Test Name. That is:

- Test suite:`tests.test_it`.
- Test name: `test_whatever`.

### What is the result that you expected?
Datadog CI Test Runs presents results for both tests with different Test Suite values that includes the test classes. For example:

- `tests.test_it.TestFoo`
- `tests.test_it.TestBar`

**Comment by albertvaka** [comment]
I agree this is an oversight and should be fixed. I'm not sure if we should add it to the test suite (`tests.test_it.TestFoo -> test_whatever`) or the test name (`tests.test_it -> TestFoo.test_whatever`), though.

**Comment by albertvaka** [comment]
Oh, I realize you are using `unittest`. However, our test instrumentation only supports `pytest`, so even though you can ruin `unittest` tests using `pytest`, classes that are specific to `unittest` like `TestCase` will not be instrumented.

Support for `unittest` is in the roadmap, but we don't have an ETA at the moment.

**Comment by apizarro-paypal** [comment]
@albertvaka Thanks for the quick reply! This behavior is not specific to `unittest` test cases, though. In my case, my test suite includes both `unitest` and Pytest test case classes. The latter are in the form:

```
class TestFoo(object):
    def test_whatever(self):
        pass


class TestBar(object):
    def test_whatever(self):
        pass
```

And in those cases, once again, the class name is nowhere to be seen in "Test suite" and "Test name".

**Comment by jirimakarius** [comment]
I had the same issue for pytest classes and UnitTest classes. 

```python
@parameterized_class(("param",), [(True,), (False,)])
class DDTraceParameterizedTestCase(SimpleTestCase):
    param = False

    def test_1(self, *args, **kwargs):
        assert self.param


class TestDDTrace:
    def test_2(self):
        assert True


class TestDDTrace2:
    def test_2(self):
        raise False
```

They are detected as same test, which was rerun, in Datadog Test runs.

I have patched span naming to "fix" it by:
```python
@pytest.hookimpl(trylast=True, hookwrapper=True)
def pytest_runtest_protocol(item, nextitem):
    if hasattr(item, "cls") and hasattr(item, "_datadog_span"):
        span = item._datadog_span
        span.set_tag_str("test.name", f"{item.cls.__name__}.{item.name}")

    yield
```
This edits span name to also include classname and tests are now separate runs.

**Comment by romainkomorndatadog** [comment]
@jirimakarius , I really appreciate the input here (and for testing the xdist loading options as well).

This issue wasn't on my radar but I'll try and get it done soon. Since it involves changing values for test names, I'll likely make it an opt-in feature to avoid breaking backwards compatibility.

One question (because I've been thinking that the suite name and module name as we have them are a little odd because the test module name is sort of like a package name, stopping short of the actual module name): what do you think of this alternative? 
* reflecting the filename name in test module
* setting the suite name to the class name (or a placeholder value like `module_level` if it's a module-level function)
* leaving test as is

So for example, a `tests/my_dir/my_module.py` file with contents
```
class MyTestClass:
    def test_1():
        assert True
```
would have:
* module `tests.my_dir.my_module`
* suite: `MyTestClass`
* test: `test_1`

@apizarro-paypal , I know it's been a while, but it'd be good to have your opinion too.

TBH, I'm a bit on the fence about the idea, so I'm testing the waters to see what people think.

**Comment by romainkomorndatadog** [comment]
I like the idea of making it pluggable, so I think I'll take the route of making minimal modifications to the existing structure (by adding an option to include the class name in the test name when using classes), and registering some hooks for the identification functions (I think that's the most Pytest-native way of going about this).

**Comment by romainkomorndatadog** [comment]
#7225 and the associated backports should get this "fix" released in `1.20`, `2.0` and `2.1` patch releases, probably next week, I'd expect.

#7221 should also merge soon and be released in `2.2.0` (unless I lose the race with the release train).

**Comment by romainkomorndatadog** [comment]
The fix patch (#7225) shipped in `ddtrace` versions [1.20.6](https://github.com/DataDog/dd-trace-py/releases/tag/v1.20.6) , [2.0.4](https://github.com/DataDog/dd-trace-py/releases/tag/v2.0.4), and [2.1.2](https://github.com/DataDog/dd-trace-py/releases/tag/v2.1.2) .

`--ddtrace-include-class-name` is available from those releases to include the class name in the test name for class-based tests.

I tested it with `1.20.6` and the [Flask repo](https://github.com/pallets/flask) and it worked, so I'm going to close out this issue:
 
![image](https://github.com/DataDog/dd-trace-py/assets/136473744/f5a4921d-db35-4e21-a422-eabb36a51302)
